src/detect-lua-extensions.c

```diff
@@ -414,6 +415,90 @@ static int LuaSetFlowint(lua_State *luastate)
     return 0;
 }
 
+struct LuaDataset {
+    Dataset *set;
+};
+
+static int LuaDatasetGC(lua_State *luastate)
+{
+    SCLogDebug("GC running");
+    struct LuaDataset *s = (struct LuaDataset *)lua_touserdata(luastate, 1);
+    SCLogDebug("deref %s", s->set->name);
+    s->set = NULL;
+    return 0;
+}
+
+static int LuaDatasetGetRef(lua_State *luastate)
+{
+    if (!lua_isstring(luastate, 1)) {
+        LUA_ERROR("1st arg is not a string");
+    }
+
+    const char *name = lua_tostring(luastate, 1);
+    if (name == NULL) {
+        LUA_ERROR("null string");
+    }
+
+    Dataset *dataset = DatasetFind(name, DATASET_TYPE_STRING);
+    if (dataset == NULL) {
+        LUA_ERROR("dataset not found");
+    }
+
+    struct LuaDataset *s = (struct LuaDataset *)lua_newuserdata(luastate, sizeof(*s));
+    if (s == NULL) {
+        LUA_ERROR("failed to get userdata");
+    }
+    s->set = dataset;
+
+    // TODO not sure why we need to do this here. Would expect it to be
+    // a registration time thing.
+    lua_newtable(luastate);
+    lua_pushcfunction(luastate, LuaDatasetGC);
+    lua_setfield(luastate, -2, "__gc");
+    lua_setmetatable(luastate, -2);
+    return 1;
+}
+
+static int LuaDatasetAdd(lua_State *luastate)
```

Comment:
Related to this function: https://github.com/OISF/suricata-verify/pull/2120#discussion_r1838468205

Reply:
addressed in https://github.com/OISF/suricata/pull/12121